messenger/http_transporter.go

```diff
@@ -482,65 +488,87 @@ func (t *HTTPTransporter) messageDecoder(w http.ResponseWriter, r *http.Request)
-	for {
-		r, ok := <-decoder.Requests()
-		if !ok {
-			// Channel closed, break out of loop
-			break
+	t.processRequests(from, decoder.Requests())
+
+	// log an error if there's one waiting, otherwise move on
+	select {
+	case err, ok := <-decoder.Err():
+		if ok {
+			log.Errorf("failed to decode HTTP message: %v", err)
 		}
-		if func() bool {
-			// regardless of whether we write a Response we must close this chan
-			defer close(r.response)
+	default:
+	}
+}
 
-			//TODO(jdef) this is probably inefficient given the current implementation of the
-			// decoder: no need to make another copy of data that's already competely buffered
-			data, err := ioutil.ReadAll(r.Body)
-			if err != nil {
-				// this is unlikely given the current implementation of the decoder:
-				// the body has been completely buffered in memory already
-				log.Errorf("failed to read HTTP body: %v\n", err)
-				return true
+func (t *HTTPTransporter) processRequests(from *upid.UPID, incoming <-chan *Request) {
+	for {
+		select {
+		case r, ok := <-incoming:
+			if !ok {
+				// Channel closed, break out of loop
+				return
 			}
-			log.V(2).Infof("Receiving %s %v from %v, length %v\n", r.Method, r.URL, from, len(data))
-			m := &Message{
-				UPID:  from,
-				Name:  extractNameFromRequestURI(r.RequestURI),
-				Bytes: data,
+			if !t.processOneRequest(from, r) {
+				return
 			}
+		case <-t.shouldQuit:
+			return
+		}
+	}
+}
 
-			keepGoing := true
-			select {
-			case <-t.shouldQuit:
-				keepGoing = false
-			case t.messageQueue <- m:
-			}
+func (t *HTTPTransporter) processOneRequest(from *upid.UPID, request *Request) (keepGoing bool) {
+	// regardless of whether we write a Response we must close this chan
+	defer close(request.response)
+	keepGoing = true
 
-			// if user-agent != libprocess then we need to send a response
-			if _, ok := parseLibprocessAgent(r.Header); !ok {
-				log.V(2).Infof("not libprocess agent, sending a 202")
-				r.response <- Response{
-					code:   202,
-					reason: "Accepted",
-				}
-			}
+	//TODO(jdef) this is probably inefficient given the current implementation of the
+	// decoder: no need to make another copy of data that's already competely buffered
+	data, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		// this is unlikely given the current implementation of the decoder:
+		// the body has been completely buffered in memory already
+		log.Errorf("failed to read HTTP body: %v", err)
+		return
+	}
+	log.V(2).Infof("Receiving %q %v from %v, length %v", request.Method, request.URL, from, len(data))
+	m := &Message{
+		UPID:  from,
+		Name:  extractNameFromRequestURI(request.RequestURI),
+		Bytes: data,
+	}
 
-			return keepGoing
-		}() {
-			continue
+	// deterministic behavior and output..
+	select {
+	case <-t.shouldQuit:
+		keepGoing = false
+		select {
+		case t.messageQueue <- m:
+		default:
+		}
+	case t.messageQueue <- m:
+		select {
+		case <-t.shouldQuit:
```

Comment:
deterministic behavior

On Thu, Sep 24, 2015 at 5:40 AM, Tomás Senart <notifications@github.com>
wrote:

> In messenger/http_transporter.go
> <https://github.com/mesos/mesos-go/pull/189#discussion_r40301299>:
>
> >
> > -			return keepGoing
> > -		}() {
> > -			continue
> > +	// deterministic behavior and output..
> > +	select {
> > +	case <-t.shouldQuit:
> > +		keepGoing = false
> > +		select {
> > +		case t.messageQueue <- m:
> > +		default:
> > +		}
> > +	case t.messageQueue <- m:
> > +		select {
> > +		case <-t.shouldQuit:
>
> Since we're already receiving from t.shouldQuit in processRequests, why
> do we have to do it here too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesos/mesos-go/pull/189/files#r40301299>.
>